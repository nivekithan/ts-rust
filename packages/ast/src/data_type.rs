use indexmap::IndexMap;

#[derive(PartialEq, Debug, Clone)]
pub enum DataType {
    Float,
    String,
    Boolean,

    Void, // only permitted as a function return type

    Unknown,
    ArrayType {
        base_type: Box<DataType>,
    },
    ObjectType {
        entries: IndexMap<String, DataType>,
    },
    FunctionType {
        return_type: Box<DataType>,
        arguments: IndexMap<String, DataType>,
    },
}
