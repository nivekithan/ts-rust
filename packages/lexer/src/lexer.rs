use crate::token::LiteralKind;

use super::token::KeywordKind;
use core::panic;
use std::str::Chars;

use super::token::Token;

pub(crate) struct Lexer<'a> {
    content: Chars<'a>,
    cur_char: Option<char>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(input: &'a str) -> Lexer<'a> {
        let content = input.chars();

        let mut lexer = Lexer {
            content,
            cur_char: None,
        };

        lexer.next();

        return lexer;
    }

    pub(crate) fn next_token(&mut self) -> Token {
        self.eat_whitespace();

        let cur_char = self.cur_char;

        match cur_char {
            None => Token::Eof,
            Some(char) => {
                use Token::*;

                if char == ';' {
                    self.next(); // consumes ;
                    return SemiColon;
                } else if char == ':' {
                    self.next(); // consumes :
                    return Colon;
                } else if char == '=' {
                    self.next(); // consumes =
                    return Assign;
                } else if char == '{' {
                    self.next(); // consumes {
                    return AngleOpenBracket;
                } else if char == '}' {
                    self.next(); // consumes }
                    return AngleCloseBracket;
                } else if char == '(' {
                    self.next(); // consumes (
                    return CurveOpenBracket;
                } else if char == ')' {
                    self.next(); // consumes )
                    return CurveCloseBracket;
                } else if char == ',' {
                    self.next(); // consumes ,
                    return Comma;
                } else if char == '!' {
                    self.next(); // consumes !
                    return Bang;
                } else if char == '+' {
                    self.next(); // consumes +
                    return Plus;
                } else if char == '-' {
                    self.next(); // consumes -
                    return Minus;
                } else if char == '*' {
                    self.next(); // consumes +
                    return Star;
                } else if char == '/' {
                    self.next(); // consumes /
                    return Slash;
                } else if char == '|' {
                    self.next(); // consumes |
                    return VerticalBar;
                } else if char == '^' {
                    self.next(); // consumes ^
                    return Caret;
                } else if char == '&' {
                    self.next(); // consumes &
                    return Ampersand;
                } else if char == '\'' {
                    let string_name = self.read_string('\'');
                    return Literal(LiteralKind::String { name: string_name });
                } else if char == '\"' {
                    let string_name = self.read_string('\"');
                    return Literal(LiteralKind::String { name: string_name });
                } else if char == '`' {
                    let string_name = self.read_string('`');
                    return Literal(LiteralKind::String { name: string_name });
                } else {
                    if is_letter(&char) {
                        let ident_name = self.read_identifier();
                        let may_be_keyword = is_keyword(&ident_name);

                        match may_be_keyword {
                            IsKeyword::Yes(kind) => return Keyword(kind),
                            IsKeyword::No => return Ident { name: ident_name },
                        }
                    } else if is_digit(&char) {
                        let digit_name = self.read_digit();
                        let digit_value: f64 = digit_name.parse().unwrap();

                        return Literal(LiteralKind::Float {
                            name: digit_name,
                            value: digit_value,
                        });
                    }
                    self.next(); // consumes Illegal
                    return Illegal;
                }
            }
        }
    }

    fn next(&mut self) -> Option<char> {
        let next_cur = self.content.next();
        self.cur_char = next_cur;

        return next_cur;
    }

    // If cur_char is whitespace it wil eat char until cur_char is not whitespace
    fn eat_whitespace(&mut self) {
        let cur_char = self.cur_char;

        if let Some(mut c) = cur_char {
            while is_whitespace(&c) {
                let next_char = self.next();
                if let Some(ch) = next_char {
                    c = ch
                } else {
                    break;
                }
            }
        }
    }

    // After executing this function the cur_char wont be a letter
    fn read_identifier(&mut self) -> String {
        let mut ident_name = self.cur_char.expect("unreachable").to_string();
        let next_cur = self.next();

        match next_cur {
            None => return ident_name,
            Some(mut c) => {
                while is_letter(&c) {
                    ident_name.push(c);

                    let next_cur = self.next();

                    match next_cur {
                        None => return ident_name,
                        Some(ch) => c = ch,
                    }
                }

                return ident_name;
            }
        }
    }
    // After executing this function the cur_char wont be a digit
    fn read_digit(&mut self) -> String {
        let mut digit_name = self.cur_char.expect("unreachable").to_string();

        let next_cur = self.next();

        match next_cur {
            None => return digit_name,
            Some(mut c) => {
                while is_digit(&c) {
                    digit_name.push(c);

                    let next_cur = self.next();

                    match next_cur {
                        None => return digit_name,
                        Some(ch) => c = ch,
                    }
                }

                return digit_name;
            }
        }
    }
    // Assumes the cur_char is starting char of string_literal
    // Ex: ' " `
    // It will end after consuming the end char that
    // cur_char wont be end_char
    fn read_string(&mut self, end_char: char) -> String {
        let mut string_name = String::new();
        loop {
            let next_char = self.next();

            match next_char {
                None => panic!("Lexer error expected {} before end of file", end_char),
                Some(ch) => {
                    if ch == end_char {
                        self.next(); // consumes end_char
                        break;
                    } else {
                        string_name.push(ch)
                    }
                }
            }
        }

        return string_name;
    }
}

// From rust source code
fn is_whitespace(c: &char) -> bool {
    return matches!(
        c,
        '\u{0009}'   // \t
        | '\u{000A}' // \n
        | '\u{000B}' // vertical tab
        | '\u{000C}' // form feed
        | '\u{000D}' // \r
        | '\u{0020}' // space

        // NEXT LINE from latin1
        | '\u{0085}'

        // Bidi markers
        | '\u{200E}' // LEFT-TO-RIGHT MARK
        | '\u{200F}' // RIGHT-TO-LEFT MARK

        // Dedicated whitespace characters from Unicode
        | '\u{2028}' // LINE SEPARATOR
        | '\u{2029}' // PARAGRAPH SEPARATOR)
    );
}

fn is_letter(c: &char) -> bool {
    return matches!(
        c,
         'a'..='z'
        | 'A'..='Z'
        | '_'
    );
}

fn is_digit(c: &char) -> bool {
    return matches!(c, '0'..='9');
}

pub enum IsKeyword {
    Yes(KeywordKind),
    No,
}

fn is_keyword(word: &String) -> IsKeyword {
    if word == "const" {
        return IsKeyword::Yes(KeywordKind::Const);
    } else if word == "true" {
        return IsKeyword::Yes(KeywordKind::True);
    } else if word == "false" {
        return IsKeyword::Yes(KeywordKind::False);
    } else if word == "let" {
        return IsKeyword::Yes(KeywordKind::Let);
    } else if word == "if" {
        return IsKeyword::Yes(KeywordKind::If);
    } else {
        return IsKeyword::No;
    }
}
