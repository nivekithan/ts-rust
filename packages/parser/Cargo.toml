[package]
name = "parser"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer = { path = "../lexer"}
ast = {path = "../ast"}
indexmap = "1.7.0"
