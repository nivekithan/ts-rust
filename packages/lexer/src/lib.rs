mod lexer;
mod token;

use crate::lexer::Lexer;
use crate::token::Token;

#[cfg(test)]
mod test;

pub fn convert_to_token(input: &str) -> Vec<Token> {
    let mut lexer = Lexer::new(input);
    let mut tokens: Vec<Token> = Vec::new();

    loop {
        let token = lexer.next_token();

        tokens.push(token);

        if let Token::Eof = token {
            break;
        }
    }

    return tokens;
}
