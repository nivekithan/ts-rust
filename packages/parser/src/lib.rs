mod parse_block;
mod parse_expression;
mod parse_type_expression;
mod parser;
mod symbol_table;
mod utils;

#[cfg(test)]
mod tests;

use crate::{parser::Parser, symbol_table::SymbolContext};
use ast::Ast;
use lexer::token::Token;

pub fn convert_to_ast(input: Vec<Token>) -> Vec<Ast> {
    let mut parser = Parser::new(&input);
    let mut asts: Vec<Ast> = vec![];
    let mut context = SymbolContext::new_empty_context();

    while parser.get_cur_token().unwrap() != &Token::Eof {
        let next_ast = parser.next_ast(&mut context);
        asts.push(next_ast);
    }

    return asts;
}

#[cfg(test)]
mod test {
    use std::collections::HashMap;

    use ast::{
        data_type::DataType,
        declaration::{VariableDeclarationKind},
        expression::Expression,
        Ast,
    };
    use indexmap::indexmap;
    use lexer::convert_to_token;

    use crate::convert_to_ast;

    #[test]
    fn test_2() {
        let input = "
        const x = {a : 1, b : 2};
        const y = x.a;";

        let mut exp_hash_map: HashMap<String, Expression> = HashMap::new();

        exp_hash_map.insert(
            "a".to_string(),
            Expression::FloatLiteralExp {
                name: "1".to_string(),
                value: 1.0,
            },
        );
        exp_hash_map.insert(
            "b".to_string(),
            Expression::FloatLiteralExp {
                name: "2".to_string(),
                value: 2.0,
            },
        );

        let expected_output: Vec<Ast> = vec![
            Ast::new_variable_declaration(
                "x_",
                Expression::ObjectLiteral {
                    expression: exp_hash_map,
                    data_type: DataType::ObjectType {
                        entries: indexmap! {"a".to_string() => DataType::Float, "b".to_string() => DataType::Float},
                    },
                },
                VariableDeclarationKind::Const,
            ),
            Ast::new_variable_declaration(
                "y_",
                Expression::DotMemberAccess {
                    container: Box::new(Expression::IdentExp {
                        data_type: DataType::ObjectType {
                            entries: indexmap! {"a".to_string() => DataType::Float, "b".to_string() => DataType::Float},
                        },
                        name: "x_".to_string(),
                    }),
                    argument: "a".to_string(),
                },
                VariableDeclarationKind::Const,
            ),
        ];

        let actual_output = convert_to_ast(convert_to_token(input));

        assert_eq!(expected_output, actual_output);
    }

    #[test]
    fn test_3() {
        let input = "
        function foo(x : number) : number {
            const x = 1;
        }";

        let expected_output = vec![Ast::new_function_declaration(
            indexmap! {"x".to_string() => DataType::Float},
            Box::new(vec![Ast::new_variable_declaration(
                "x_",
                Expression::FloatLiteralExp {
                    name: "1".to_string(),
                    value: 1.0,
                },
                VariableDeclarationKind::Const,
            )]),
            "foo_".to_string(),
            DataType::Float,
        )];

        let actual_output = convert_to_ast(convert_to_token(input));

        assert_eq!(expected_output, actual_output);
    }
}
