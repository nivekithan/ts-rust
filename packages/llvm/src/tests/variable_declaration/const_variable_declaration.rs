use ast::data_type::DataType;
use lexer::convert_to_token;
use parser::convert_to_ast;
use test_utils::{DatatypeOrFn, ExpressionTest};

use crate::write_llvm_ir;

#[test]
fn test_all_const() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = {}", main_strings[0]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![(
            DatatypeOrFn::Fn(Box::new(|_| {
                return true;
            })),
            vec!["_x".to_string()],
        )],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn explicit_float_type_const() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x : number = {}", main_strings[0]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![(
            DatatypeOrFn::DataType(DataType::Float),
            vec!["_x".to_string()],
        )],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn explicit_boolean_type_const() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x : boolean = {}", main_strings[0]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![(
            DatatypeOrFn::DataType(DataType::Boolean),
            vec!["_x".to_string()],
        )],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn test_const_unary_bang() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = !({})", main_strings[0]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![(
            DatatypeOrFn::DataType(DataType::Boolean),
            vec!["_x".to_string()],
        )],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn unary_plus() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = +({})", main_strings[0]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![(
            DatatypeOrFn::DataType(DataType::Float),
            vec!["_x".to_string()],
        )],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn unary_minus() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = -({})", main_strings[0]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![(
            DatatypeOrFn::DataType(DataType::Float),
            vec!["_x".to_string()],
        )],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_float_plus() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) + ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_float_minus() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) - ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_float_star() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) * ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_float_slash() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) / ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_strict_equality_bool() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) === ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Boolean),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Boolean),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_strict_equality_float() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) === ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_strict_not_equal_bool() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) !== ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Boolean),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Boolean),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_strict_not_equal_float() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) !== ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_less_than() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) < ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_less_than_or_equal() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) <= ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_greater_than() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) > ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}

#[test]
fn binary_greater_than_or_equal() {
    let exp_test = ExpressionTest {
        generate_input: Box::new(|ast_strings, main_strings| {
            let ast_input = "\n".to_string() + ast_strings.join("\n").as_str();

            let main_input = format!("const x  = ({}) >= ({})", main_strings[0], main_strings[1]);

            let input = ast_input + main_input.as_str();

            return input;
        }),

        expressions_data_type: vec![
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["_x".to_string()],
            ),
            (
                DatatypeOrFn::DataType(DataType::Float),
                vec!["__y".to_string()],
            ),
        ],

        test: Box::new(|_, _, input| {
            let output = write_llvm_ir(convert_to_ast(convert_to_token(input.as_str())));

            insta::assert_snapshot!(input, output);
        }),
    };

    exp_test.test();
}


#[test]
fn literal_object_declaration() {
    let input = "

    const x = {a : 5};
    const y = x.a;
    ";

    let output = write_llvm_ir(convert_to_ast(convert_to_token(input)));

    insta::assert_snapshot!(input, output);
}


#[test]
fn literal_object_nested_object_declaration() {
    let input = "

    const x = {a : {a : 5}};
    const y = x.a;
    const z = y.a;
    ";

    let output = write_llvm_ir(convert_to_ast(convert_to_token(input)));
    
    insta::assert_snapshot!(input, output);
}