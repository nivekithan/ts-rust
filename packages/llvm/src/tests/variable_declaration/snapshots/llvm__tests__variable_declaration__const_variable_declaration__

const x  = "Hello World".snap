---
source: packages/llvm/src/tests/variable_declaration/const_variable_declaration.rs
expression: output

---
; ModuleID = 'main'
source_filename = "main"

define void @main() {
entry:
  %"x|_|" = alloca [11 x i8], align 1
  %"0" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 0
  store i8 72, i8* %"0", align 1
  %"1" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 1
  store i8 101, i8* %"1", align 1
  %"2" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 2
  store i8 108, i8* %"2", align 1
  %"3" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 3
  store i8 108, i8* %"3", align 1
  %"4" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 4
  store i8 111, i8* %"4", align 1
  %"5" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 5
  store i8 32, i8* %"5", align 1
  %"6" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 6
  store i8 87, i8* %"6", align 1
  %"7" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 7
  store i8 111, i8* %"7", align 1
  %"8" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 8
  store i8 114, i8* %"8", align 1
  %"9" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 9
  store i8 108, i8* %"9", align 1
  %"10" = getelementptr [11 x i8], [11 x i8]* %"x|_|", i64 0, i64 10
  store i8 100, i8* %"10", align 1
  ret void
}

